File: project RAD/ribbitproblem/sol.py
def jumpedOn(head: SinglyLinkedListNode) -> bool:
    # 1. Convert Linked List to Array
    arr = convert_list_to_array(head)
    n = len(arr)
    
    if n == 0:
        return True # Or handle as per problem spec for empty list

    # Set to track visited indices
    visited = set()
    
    # Start at index 0
    current_index = 0
    
    while current_index not in visited:
        # Add current index to visited set
        visited.add(current_index)
        
        # Check if we have successfully visited all N nodes
        if len(visited) == n:
            return True # All nodes visited

        # Get jump value (v) and calculate the next index
        jump_value = arr[current_index]
        
        # Apply the jump rule: (i + v) mod n
        # Note: In Python, % handles negative numbers correctly for cyclic indices
        next_index = (current_index + jump_value) % n 
        
        # Move to the next index
        current_index = next_index
        
    # If the loop breaks because current_index is already in visited,
    # it means a cycle was found before all nodes were visited.
    return False

# Time Complexity: O(N) for conversion and O(N) for the single-pass cycle check. Total O(N).
# Space Complexity: O(N) to store the array and the visited set.

File: project RAD/ribbitproblem/sol.java
class Solution {

    public static class ListNode {
        int val;
        ListNode next;
        ListNode(int val) { this.val = val; }
    }

    public boolean jumpedOn(ListNode head) {
        // Step 1: Convert Linked List to Array
        int[] arr = toArray(head);
        int n = arr.length;

        if (n == 0) return true;

        boolean[] visited = new boolean[n];

        int current = 0;
        int visitCount = 0;

        while (!visited[current]) {
            visited[current] = true;
            visitCount++;

            if (visitCount == n) {
                return true;  // visited all nodes
            }

            int jump = arr[current];
            current = mod(current + jump, n);
        }

        // If we reached a visited index early → bad cycle
        return false;
    }

    // Helper: Convert linked list to array
    private int[] toArray(ListNode head) {
        java.util.ArrayList<Integer> list = new java.util.ArrayList<>();
        ListNode curr = head;
        while (curr != null) {
            list.add(curr.val);
            curr = curr.next;
        }
        // convert List<Integer> → int[]
        return list.stream().mapToInt(i -> i).toArray();
    }

    // Proper modulo for negative values
    private int mod(int a, int n) {
        int r = a % n;
        return r < 0 ? r + n : r;
    }
}
